File: packages/ts/src/guarded-generate.ts
/**
 * guardedGenerate: Async wrapper that calls a generate function, validates/repairs
 * output, and retries with a correction prompt on failure.
 *
 * F4 compliance: AbortController per-call timeout (default 30s).
 * MAX_RETRIES=3 hard cap.
 */
import type { ZodType } from "zod";
import type { Format, GuardedGenerateResult, RepairReport } from "./results.js";
import { GuardedGenerationError } from "./errors.js";
import { MAX_RETRIES, DEFAULT_TIMEOUT_MS } from "./constants.js";
import { repair } from "./orchestrator.js";
import { retryPrompt } from "./retry-prompt.js";
import { zodAdapter } from "./adapters/zod.js";

export interface Message {
  role: "user" | "assistant" | "system";
  content: string;
}

export interface GuardedGenerateOptions<T> {
  generate: (prompt: string, history?: Message[]) => Promise<string>;
  prompt: string;
  format?: Format;
  schema?: ZodType<T>;
  maxRetries?: number;
  timeoutMs?: number;
  signal?: AbortSignal;
}

const withTimeout = <T>(
  promise: Promise<T>,
  ms: number,
  signal?: AbortSignal,
): Promise<T> => {
  return new Promise<T>((resolve, reject) => {
    if (signal?.aborted) {
      reject(new Error("AbortSignal already aborted"));
      return;
    }
    const timer = setTimeout(() => reject(new Error(`Timeout after ${ms}ms`)), ms);
    const onAbort = () => {
      clearTimeout(timer);
      reject(new Error("Aborted"));
    };
    signal?.addEventListener("abort", onAbort, { once: true });
    promise.then(
      (val) => { clearTimeout(timer); signal?.removeEventListener("abort", onAbort); resolve(val); },
      (err: unknown) => { clearTimeout(timer); signal?.removeEventListener("abort", onAbort); reject(err); },
    );
  });
};

export const guardedGenerate = async <T = unknown>(
  opts: GuardedGenerateOptions<T>,
): Promise<GuardedGenerateResult<T>> => {
  const {
    generate,
    prompt,
    format = "auto",
    schema,
    maxRetries = MAX_RETRIES,
    timeoutMs = DEFAULT_TIMEOUT_MS,
    signal,
  } = opts;

  const validator = schema ? zodAdapter(schema).validate : undefined;
  const history: Message[] = [];
  let lastError: unknown;
  let lastResponse = "";

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    if (signal?.aborted) {
      throw new GuardedGenerationError("Aborted by signal", attempt, lastError);
    }

    const currentPrompt =
      attempt === 0
        ? prompt
        : retryPrompt({
            previousResponse: lastResponse,
            errors: lastError instanceof Error ? [lastError.message] : [String(lastError)],
            history,
          }).prompt;

    let raw: string;
    try {
      raw = await withTimeout(generate(currentPrompt, history.length > 0 ? history : undefined), timeoutMs, signal);
    } catch (e) {
      throw new GuardedGenerationError(
        `Generation failed on attempt ${attempt}: ${String(e)}`,
        attempt,
        e,
      );
    }

    lastResponse = raw;
    history.push({ role: "user", content: currentPrompt });
    history.push({ role: "assistant", content: raw });

    try {
      const repairResult = repair<T>(raw, format, validator);
      const report: RepairReport = {
        strategiesApplied: repairResult.strategiesApplied,
        retries: attempt,
        finalValid: true,
        pass: repairResult.repaired ? "B" : "A",
      };
      return { data: repairResult.data as T, report };
    } catch (e) {
      lastError = e;
    }
  }

  throw new GuardedGenerationError(
    `Exhausted ${maxRetries} retries without valid output`,
    maxRetries,
    lastError,
  );
};

File: packages/python/src/output_guard/guarded_generate.py
"""guarded_generate: Async wrapper around a generate callable with repair + retry.

Mirror of packages/ts/src/guarded-generate.ts.

F4 compliance: ``asyncio.timeout`` per-call (default 30s). ``MAX_RETRIES=3`` hard cap.
"""

from __future__ import annotations

import asyncio
import inspect
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import Any

from pydantic import BaseModel

from .adapters.pydantic_ import SchemaAdapter, pydantic_adapter
from .constants import DEFAULT_TIMEOUT_MS, MAX_RETRIES
from .errors import GuardedGenerationError
from .orchestrator import repair
from .results import Format, GuardedGenerateResult, RepairReport
from .retry_prompt import Message, RetryPromptOptions, retry_prompt

GenerateCallable = Callable[..., Awaitable[str]] | Callable[..., str]


@dataclass(slots=True)
class GuardedGenerateOptions:
    generate: GenerateCallable
    prompt: str
    format: Format = "auto"
    schema: SchemaAdapter[Any] | type[BaseModel] | None = None
    max_retries: int = MAX_RETRIES
    timeout_ms: int = DEFAULT_TIMEOUT_MS
    history: list[Message] = field(default_factory=list)


def _resolve_schema(
    schema: SchemaAdapter[Any] | type[BaseModel] | None,
) -> SchemaAdapter[Any] | None:
    if schema is None:
        return None
    if isinstance(schema, SchemaAdapter):
        return schema
    return pydantic_adapter(schema)


async def _invoke(
    generate: GenerateCallable,
    prompt: str,
    history: list[Message] | None,
    timeout_ms: int,
) -> str:
    async def runner() -> str:
        if inspect.iscoroutinefunction(generate):
            try:
                value: Any = await generate(prompt, history)
            except TypeError:
                value = await generate(prompt)
            assert isinstance(value, str)
            return value
        result: Any = generate(prompt, history) if _accepts_history(generate) else generate(prompt)
        if inspect.isawaitable(result):
            awaited: Any = await result
            assert isinstance(awaited, str)
            return awaited
        assert isinstance(result, str)
        return result

    return await asyncio.wait_for(runner(), timeout=timeout_ms / 1000)


def _accepts_history(fn: Callable[..., Any]) -> bool:
    try:
        sig = inspect.signature(fn)
    except (TypeError, ValueError):
        return False
    return len(sig.parameters) >= 2


async def guarded_generate(opts: GuardedGenerateOptions) -> GuardedGenerateResult:
    """Call ``opts.generate``, validate/repair output, retry with correction prompt on failure."""
    adapter = _resolve_schema(opts.schema)
    history: list[Message] = list(opts.history)
    last_error: BaseException | None = None
    last_response = ""

    for attempt in range(opts.max_retries + 1):
        current_prompt = (
            opts.prompt
            if attempt == 0
            else retry_prompt(
                RetryPromptOptions(
                    previous_response=last_response,
                    errors=[str(last_error)] if last_error is not None else [],
                    history=history,
                )
            ).prompt
        )

        try:
            raw = await _invoke(opts.generate, current_prompt, history or None, opts.timeout_ms)
        except (TimeoutError, Exception) as e:
            raise GuardedGenerationError(
                f"Generation failed on attempt {attempt}: {e}", attempt, e
            ) from e

        last_response = raw
        history.append(Message(role="user", content=current_prompt))
        history.append(Message(role="assistant", content=raw))

        try:
            repair_result = repair(raw, opts.format, adapter)
            report = RepairReport(
                strategies_applied=list(repair_result.strategies_applied),
                retries=attempt,
                final_valid=True,
                pass_=repair_result.pass_,
            )
            return GuardedGenerateResult(data=repair_result.data, report=report)
        except Exception as e:
            last_error = e

    raise GuardedGenerationError(
        f"Exhausted {opts.max_retries} retries without valid output",
        opts.max_retries,
        last_error,
    )
